### PR TITLE
Feature: MicroSplit Dataset

### PR DESCRIPTION
**Pull request by melisande-c**
## Disclaimer

- [ ] I am an AI agent.
- [x] I have used AI and I thoroughly reviewed every line.
- [ ] I have not used AI extensively.

## Description

Seeing as this is a very big PR there are a couple options:
1. Some brave soul can review this as it is. A lot of things have changed so maybe some things will fall through the cracks.
2. I keep this PR open as reference (and for current ongoing work), but I split the work up into smaller more reviewable PRs, maybe to a `dev/microsplit-dataset` branch. This will take longer.

> [!NOTE]  
> **tldr**: Very big PR for the introduction of the MicroSplit dataset... Includes a fairly big refactoring of the existing dataset.

### Background - why do we need this PR?

In the end I decided to make MicroSplit compatible with the existing dataset to reuse the normalisation and and augmentation logic. To do this I had to introduce a new component which I named `PatchConstructor`, (confusingly there was previously a `PatchConstructor` class for something else that I have now deleted, but I have run out of creativity for naming things). The `PatchConstructor` class wraps the the `PatchingStrategy` and `PatchExtractor` and handles the production of input and target patches, hiding this implementation logic from the dataset. The reason to do this is that MicroSplit has some more complex patch construction logic for creating synthetic inputs for training.

Additionally, LC patch construction was previously happening inside the `PatchExtractor` (using the old `PatchConstructor` class), but this mechanism was a bit opaque so I have moved it to the new `PatchConstructor` which hopefully makes it easier to follow the logic.

In turn this refactoring lead to refactoring the normalization statistics calculation logic. It previously first iterated through the input data and then the target data. But now both the input and target patches are produced together, necessarily for the synthetic inputs.

For the data config a new `MicroSplitDataConfig` has been added. This is for the additional parameters for LC patch construction: "multiscale_count" and "padding_mode". And synthetic input parameters: "alpha_ranges" and "uncorrelated_patch_prob".

Dataset factory functions have also been added.

There is also a `UncorrelatedPatchSpecs` which stores the patch specs of other channels in an additional field. The resulting `ImageRegionData` objects have the additional sources added to the `"additional_metadata"` field.

### Implementation - how did you implement the changes?

There are 4 modes that the MicroSplit dataset needs to operate in, 3 training modes that are described in the paper, and prediction mode:

  - Training mode I: For multiplexed target images, but no real input is available so it is synthetically created by summing the target channels.
  - Training mode II: For target channels that are separately acquired, the different channels are not registered and may even be different shapes, e.g. puncta experiment.
  - Training mode III: For multiplexed target images, with a real input of the superimposed structures that was also acquired at the time the target was acquired.
  - Inference/prediction mode: There is only input images.

For these 4 modes I have created 4 different `PatchConstructor` classes called `MsT1PatchConstructor`, `MsT2PatchConstructor`, `MsT3PatchConstructor` and `MsPredPatchConstructor`. Pls help me think of better names 

To help make the code paths clear in the factory functions there are `MicroSplitMultiplexedTargetData`, `MicroSplitSeparateTargetData` and `MicroSplitPairedData`, data classes.

### New features or files

- `MicroSplitDataConfig`

- `Normalize.target_needs_computation`

- dataclasses for MicroSplit training data:
  - `MicroSplitMultiplexedTargetData`
  - `MicroSplitSeparateTargetData`
  - `MicroSplitPairedData`

- `create_microsplit_dataset` factory function
- `create_microsplit_pred_dataset` factory function

- `PatchConstructor` protocol
- `BasicPatchConstructor`
- `MsT1PatchConstructor`
- `MsT2PatchConstructor`
- `MsT3PatchConstructor`
- `MsPredPatchConstructor`

- `UncorrelatedPatchSpecs`

### Modified features or files

- `CareamicsDataset`

- `PatchExtractor` (removed patch constructor mechanic)

- `_compute_mean_std`
- `_compute_min_max`
- `_compute_quantiles`
- `resolve_normalization_config`

### Removed features or files

- `careamics/dataset/patch_extractor/patch_construction.py` (old `PatchConstructor` class)
- `careamics/dev/microsplit_input_synth.py`

## How has this been tested?

- Modified existing tests for lateral context creation
- Tests for dataset factory for the 3 different MicroSplit training modes.
- Test the prediction datasets factory
- Test normalization stats are calculated for the 3 different training modes. (Matrix of training mode and normalization type)
- `MicroSplitDataConfig` validation tests
- Tests for each of the MicroSplit patch constructors `construct_patch` method
- Tests for uncorrelated metadata

## Related Issues

Would probably close a few, need to take a look

- Resolves #

## Breaking changes

Initialising the dataset class directly.

## Additional Notes and Examples

Training modes I & III:
```python
import nd2
import pooch

from careamics.config.data import MicroSplitDataConfig
from careamics.config.factories import create_ng_data_configuration
from careamics.dataset.factory import (
    MicroSplitMultiplexedTargetData,
    MicroSplitPairedData,
    create_microsplit_dataset,
)

# get data
DATA = pooch.create(
    path="../data",
    base_url="https://download.fht.org/jug/msplit/ht_lif24/data/",
    registry={"ht_lif24_5ms.zip": None},
)
for fname in DATA.registry:
    DATA.fetch(fname, processor=pooch.Unzip(), progressbar=True)
DATA_ROOT = DATA.path / "ht_lif24_5ms.zip.unzip/5ms"
example_path = sorted(list(DATA_ROOT.glob("**/*.nd2")))[0]
img = nd2.imread(example_path)

# create config
base_config = create_ng_data_configuration(
    "array", "SCYX", (64, 64), 1
)
config = MicroSplitDataConfig(
    **base_config.model_dump(),
    multiscale_count=3
)

# target only, synthetic input (training mode I)
dataset = create_microsplit_dataset(
    config,
    MicroSplitMultiplexedTargetData(
        target_data=[img[:, [0, 1], ...].astype("float32")],
    ),
)

# both real input and target (training mode III)
dataset = create_microsplit_dataset(
    config,
    MicroSplitPairedData(
        input_data=[img[:, [8], ...].astype("float32")],
        target_data=[img[:, [0, 1], ...].astype("float32")],
    ),
)
```

Training mode II:
```python
import pooch

from careamics.config.data import MicroSplitDataConfig
from careamics.config.factories import create_ng_data_configuration
from careamics.dataset.factory import (
    MicroSplitSeparateTargetData,
    create_microsplit_dataset,
)

# download data
DATA = pooch.create(
    path="../data/",
    base_url="https://download.fht.org/jug/msplit/ht_h23b/data/",
    registry={"ht_h23b.zip": None},
)
for fname in DATA.registry:
    DATA.fetch(fname, processor=pooch.Unzip(), progressbar=True)
DATA_ROOT = DATA.path / "ht_h23b.zip.unzip/ht_h23b"
CHANNEL_DIRS = ["foreground", "puncta"]

# create config
base_config = create_ng_data_configuration(
    "tiff", "YX", (64, 64), 1
)
config = MicroSplitDataConfig(
    **base_config.model_dump(),
    multiscale_count=1
)

# create dataset
data = MicroSplitSeparateTargetData(
    [(DATA_ROOT / channel_dir).glob("*.tiff") for channel_dir in CHANNEL_DIRS]
)
dataset = create_microsplit_dataset(
    config=config,
    data=data
)
```

For prediction:
```python
import pooch
import nd2

from careamics.config.data import MicroSplitDataConfig
from careamics.config.factories import create_ng_data_configuration
from careamics.dataset.factory import create_microsplit_pred_dataset

DATA = pooch.create(
    path="../data",
    base_url="https://download.fht.org/jug/msplit/ht_lif24/data/",
    registry={"ht_lif24_5ms.zip": None},
)
for fname in DATA.registry:
    DATA.fetch(fname, processor=pooch.Unzip(), progressbar=True)

DATA_ROOT = DATA.path / "ht_lif24_5ms.zip.unzip/5ms"

example_path = sorted(list(DATA_ROOT.glob("**/*.nd2")))[0]
img = nd2.imread(example_path)

base_config = create_ng_data_configuration(
    "array",
    "SCYX",
    (64, 64),
    1,
    normalization={
        "name": "mean_std",
        "input_means": [113.10580444335938],
        "input_stds": [19.361731573917567],
        "target_means": [101.47886657714844, 102.36434936523438],
        "target_stds": [4.220477114929124, 6.84461487895705],
    },
)
pred_config = MicroSplitDataConfig(
    **base_config.model_dump(), multiscale_count=3
).convert_mode("predicting", new_patch_size=(64, 64), overlap_size=(48, 48))

pred_dataset = create_microsplit_pred_dataset(
    pred_config, [img[:, [8], ...].astype("float32")]
)

input_patch, *_ = pred_dataset[0]
```

##

Known issues:

- Iterable tiff loading currently does not work with uncorrelated patches, for training mode II this will require a little more engineering.
- For training mode 3, it could be possible that the input has axes YX and the target has CYX, this is currently not accepted. (Currently input has to have singleton C axes)
- Should make a microsplit data config factory
- Patch filtering is currently not supported

---

**Please ensure your PR meets the following requirements:**

- [x] Code builds and passes tests locally, including doctests
- [x] New tests have been added (for bug fixes/features)
- [ ] Documentation has been updated
- [x] Pre-commit passes

---

EDIT:

Final renaming of objects:

#### Patch Constructors
`PatchConstructor` -> `PatchConstr`
`BasicPatchConstructor` -> `BasicPatchConstr`
`MsT1PatchConstructor` -> `MultiChannelTargetMsPatchConstr`
`MsT2PatchConstructor` -> `IndependentTargetsMsPatchConstr`
`MsT3PatchConstructor` -> `PairedInputTargetMsPatchConstr`
`MsPredPatchConstructor` -> `PredMsPatchConstr`

#### Data class data-wrappers
`MicroSplitMultiplexedTargetData` -> `MultiChannelTarget`
`MicroSplitSeparateTargetData` -> `IndependentTargets`
`MicroSplitPairedData` -> `PairedInputTarget`

### CODE REVIEW

File: src/careamics/dataset/factory/factory.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
     load_tiffs,
     load_zarrs,
 )
+from ..patch_constructor import BasicPatchConstructor
 from ..patch_extractor import LimitFilesPatchExtractor, PatchExtractor
 from ..patching import StratifiedPatching, create_patching
```
Comment (jdeschamps) [inline src/careamics/dataset/factory/factory.py]: can we use absolute import when the modules don't reside next to the current module?

File: src/careamics/dataset/factory/microsplit_factory.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+from careamics.config.data.microsplit_data_config import MicroSplitDataConfig
+from careamics.config.support import SupportedData
+from careamics.models.constraints import ModelConstraints
+from careamics.utils import get_logger
+
+from ..dataset import CareamicsDataset
+from ..image_stack import ImageStack
+from ..patch_constructor import PatchConstructor
+from ..patch_constructor.microsplit_patch_constructors import (
```
Comment (jdeschamps) [inline src/careamics/dataset/factory/microsplit_factory.py]: Again, not a big fan of relative imports of depth > 0 (and the code base is not doing that in general)

File: src/careamics/dataset/factory/microsplit_factory.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+    match data:
+        case MicroSplitMultiplexedTargetData(target_data):
+            _warn_unused_training_config_fields(config, "multiplexed-target mode")
+            target_extractor = init_patch_extractor(
+                patch_extractor_type, image_stack_loader, target_data, config.axes
+            )
+            patching_strategy = create_patching(
+                target_extractor.shapes, config.patching
+            )
+            patch_constructor = MsT1PatchConstructor(
```
Comment (jdeschamps) [inline src/careamics/dataset/factory/microsplit_factory.py]: Multiplexed, Separate and Paired are a good start for clearer name. Maybe they could already be applied to the patch constructors for clarity.

File: src/careamics/dataset/factory/microsplit_factory.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+                patching_strategy=patching_strategy,
+                target_extractor=target_extractor,
+                multiscale_count=config.multiscale_count,
+                padding_mode=config.padding_mode,
+                alpha_ranges=config.alpha_ranges,
+                uncorrelated_channel_prob=config.uncorrelated_channel_prob,
+                channels=config.channels,
```
Comment (jdeschamps) [inline src/careamics/dataset/factory/microsplit_factory.py:200]: I have not looked at the patch constructors yet, but it seems they could simply accept the data configuration.

Comment (jdeschamps) [reply src/careamics/dataset/factory/microsplit_factory.py:200]: I am not sure what is better for the patch constructors: a clear view of the necessary parameters or a smaller code footprint here... As far as testing goes, it is probably better to have the parameters explicit as you've done it.

File: src/careamics/dataset/normalization/statistics.py
Comment (jdeschamps) [inline src/careamics/dataset/normalization/statistics.py:1]: The new statistics behaviour should probably be written in the docs

File: src/careamics/dataset/normalization/statistics.py [hunk 5]
```diff
@@ -13,188 +13,211 @@
...
-        patch_spec = patching_strategy.get_patch_spec(idx)
-        patch = data_extractor.extract_channel_patch(
-            data_idx=patch_spec["data_idx"],
-            sample_idx=patch_spec["sample_idx"],
-            channels=channels,
-            coords=patch_spec["coords"],
-            patch_size=patch_spec["patch_size"],
-        )
+        input_patch, target_patch, _ = patch_constructor.construct_patch(idx)
+        has_target = has_target or target_patch is not None
```
Comment (jdeschamps) [inline src/careamics/dataset/normalization/statistics.py]: is `has_target` necessary? It is arguably cleaner because it helps reading the last clause, but `target_stats` should be enough to know if there is a target.

File: src/careamics/dataset/normalization/statistics.py [hunk 6]
```diff
@@ -13,188 +13,211 @@
...
-
-    means, stds = image_stats.finalize()
-    return means.tolist(), stds.tolist()
+            input_patch = input_patch.reshape(1, -1)
+            if target_patch is not None:
+                target_patch = target_patch.reshape(1, -1)
+        input_stats.update(input_patch[None, ...], sample_idx=idx)
+        if target_patch is not None:
+            target_stats.update(target_patch[None, ...], sample_idx=target_idx)
+            target_idx += 1
```
Comment (jdeschamps) [inline src/careamics/dataset/normalization/statistics.py]: would there be a case where they would be sometimes target and sometimes not? I don't understand why `target_idx` may be different from `idx`

**Review by jdeschamps — APPROVED** [review]

File: src/careamics/dataset/factory/microsplit_factory.py (hunk 4 above)
Comment (melisande-c) [reply src/careamics/dataset/factory/microsplit_factory.py:200]: Yeah I did it with explicit parameters to show which of the microsplit parameters are used for each mode, e.g `uncorrelated_channel_prob` is only used for mode 1, `alpha_ranges` are only used for modes 1&2.

File: src/careamics/dataset/normalization/statistics.py (hunk 5 above)
Comment (melisande-c) [reply src/careamics/dataset/normalization/statistics.py]: Target stats will also be None for a self supervised algorithm, but I agree this bit of code is a bit weird I'll try to improve it, this section I used an agent but edited the result

File: src/careamics/dataset/normalization/statistics.py (hunk 6 above)
Comment (melisande-c) [reply src/careamics/dataset/normalization/statistics.py]: yeah I think I missed this when reviewing the agent output, I will remove it, thanks for catching

Comment (melisande-c) [reply src/careamics/dataset/normalization/statistics.py]: I will review this whole file again bec there are some weird things happening

File: src/careamics/dataset/factory/microsplit_factory.py (hunk 3 above)
Comment (melisande-c) [reply src/careamics/dataset/factory/microsplit_factory.py]: Oh yeah I wanted to confirm that multiplexed is accurate or too narrow of a name?

File: src/careamics/dataset/normalization/statistics.py
Comment (melisande-c) [reply src/careamics/dataset/normalization/statistics.py:1]: There's not particularly new behaviour, it has just been refactored so that instead of having two sequential loops the first for the input stats and the second for the target stats, now it calculates both in a single loop

File: src/careamics/dataset/factory/factory.py (hunk 1 above)
Comment (melisande-c) [reply src/careamics/dataset/factory/factory.py]: done in [cbc903b](https://github.com/CAREamics/careamics/pull/962/commits/cbc903b3d8672a8aeec4baa031b08768a86cdec1)

File: src/careamics/dataset/factory/microsplit_factory.py (hunk 2 above)
Comment (melisande-c) [reply src/careamics/dataset/factory/microsplit_factory.py]: done in [cbc903b](https://github.com/CAREamics/careamics/pull/962/commits/cbc903b3d8672a8aeec4baa031b08768a86cdec1)

File: src/careamics/dataset/normalization/statistics.py (hunk 6 above)
Comment (melisande-c) [reply src/careamics/dataset/normalization/statistics.py]: I refactored all the functions to make them more consistent with between normalisation kinds, the stats calculation could be refactored more in the future, but this PR is already pretty big. In [f2f57c8](https://github.com/CAREamics/careamics/pull/962/commits/f2f57c87601b4c3b69ae9b6d6504af53d3032d76)

File: src/careamics/dataset/normalization/statistics.py (hunk 5 above)
Comment (melisande-c) [reply src/careamics/dataset/normalization/statistics.py]: Changed the function to take the arguments `compute_input` and `compute_target`. This means that the choice of calculating either the input or the target is independent of the other, e.g. I could supply the target stats and have only the input calculated. I accidentally did not include this feature before.

In [f2f57c8](https://github.com/CAREamics/careamics/pull/962/commits/f2f57c87601b4c3b69ae9b6d6504af53d3032d76)

File: src/careamics/dataset/factory/microsplit_factory.py (hunk 3 above)
Comment (jdeschamps) [reply src/careamics/dataset/factory/microsplit_factory.py]: I would not say it is accurate but it is better than nothing for now... Multiplexed happens when generating input in mode I and II, and in the microscope in mode III. So it may still be confusing.  

The main difference between mode I and II is that the channels are or are not spatially coupled (same sample, same ROI?). Maybe Coupled, Uncoupled and Paired?

File: src/careamics/dataset/factory/microsplit_factory.py (hunk 4 above)
Comment (jdeschamps) [reply src/careamics/dataset/factory/microsplit_factory.py:200]: Makes sense, I actually like it after thinking more about it.

File: src/careamics/dataset/normalization/statistics.py
Comment (jdeschamps) [reply src/careamics/dataset/normalization/statistics.py:1]: I can't make myself believe that we were already doing one "epoch" worth for the stats... Out of scope here, but this should be written in the docs.

File: src/careamics/dataset/normalization/statistics.py (hunk 5 above)
Comment (jdeschamps) [reply src/careamics/dataset/normalization/statistics.py]: cleaner!

File: src/careamics/dataset/factory/microsplit_factory.py (hunk 3 above)
Comment (melisande-c) [reply src/careamics/dataset/factory/microsplit_factory.py]: For mode II it is possible that only 1 channel is acquired for the use-case of removing artefacts. But the name is MultiplextedTarget, SeparateTarget and Paired, I guess the specification of "target" gets lost in the long name

Comment (melisande-c) [reply src/careamics/dataset/factory/microsplit_factory.py]: renamed after discussion in [75c351f](https://github.com/CAREamics/careamics/pull/962/commits/75c351f9ebc1875a2dc228ea2b75601cc26df0bc)
